docs/architecture/engine.md

```diff
@@ -0,0 +1,270 @@
+
+4. "Series" which are owned by shards. A series is the minimum unit that comes to mind when you think of "time series" data. Ex. The CPU level for a single host in a datacenter over a period of time could be represented as a series with id "<HOST_IDENTIFIER>.system.cpu.utilization" and a vector of tuples in the form of (TIMESTAMP, CPU_LEVEL)
+
+5. "Blocks" belong to a series and are central to M3DB's design. In order to understand blocks, we must first understand that one of M3DB's biggest strengths as a time series database (as opposed to using a more general-purpose horizontally scalable, distributed database like Cassandra) is its ability to compress time series data resulting in huge memory and disk savings. This high compression ratio is implemented via the M3TSZ algorithm, a variant of the streaming time series compression algorithm described in [Facebook's Gorilla paper](http://www.vldb.org/pvldb/vol8/p1816-teller.pdf) with a few small differences. A "block" then is simply a sealed (no longer writable) stream of compressed time series data. The compression ratio will vary depending on the workload and configuration, but with careful tuning its possible to encode data with an average of 1.4 bytes per datapoint. The compression comes with a few caveats though, namely that you cannot read individual datapoints in a compressed block. In other words, in order to read a single datapoint you must decompress the entire block up to the datapoint that you're trying to read.
+
+If M3DB kept everything in memory (and in fact, early versions of it did), than you could conceptually think of it as having a single top-level database which contained an internal map of <NAMEPSACE_ID>:<NAMESPACE_OBJECT> and each
+namespace would have an internal map of <SHARD_NUMBER>:<SHARD_OBJECT> and each shard would have an internal map of
+<SERIES_ID (in reality we use a hash)>:<SERIES_OBJECT> and each series would have an internal map of <BLOCK_START_TIME:BLOCK_OBJECT> where a block object is a thin wrapper around a compressed block of M3TSZ encoded data.
+
+In fact, even though M3DB does implement persistent storage as well, all of the in-memory datastructures described above exist in the M3DB codebase.
+
+### Persistent storage
+
+While in-memory databases can be useful (and M3DB supports operating in a memory-only mode), with large volumes of data it becomes prohibitively expensive to keep all of the data in memory. In addition, monitoring time series data often follows a "write-once, read-never" pattern where less than a few percent of all the data thats ever stored is ever read. With that type of workload, its wasteful to keep all of that data in memory when it could be persisted on disk and retrieved when required (with an appropriate caching policy for frequently accessed data).
```

Comment:
Should rephrase to include the fact you need persistence for durability before anything else?

also, most people wont' get the subtle implications of the "caching policy" reference in the last sentence. If you want to include it, could you expand on it a bit more.

Reply:
will fix